### PR TITLE
o/snapstate: add support for kernel-modules handlers

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Add handlers and backend methods that will set-up early mount units and kernel tree when installing a kernel-modules component.

### CODE REVIEW

**Comment by pedronis** [comment]
Marking blocked as I'm unsure about the underlying plan in terms of adding more handlers

**Comment by alfonsosanchezbeato** [comment]
> Marking blocked as I'm unsure about the underlying plan in terms of adding more handlers

There is a fundamental difference here that it that for kernel installs/refreshes (with or without components), you build a new kernel tree, and then you switch to it after a reboot. If you are refreshing/installing kernel-modules components for an installed kernel, you are modifying the current kernel tree and you need the modules to be available without rebooting. This means that we need two tasks for the former ("setup-kernel-snap", "remove-old-kernel-snap-setup") and only one for the latter ("setup-kernel-modules-component"). This one could be split, but that seemed to complicate things.

Also, although the low-level handler, `EnsureKernelDriversTree` is the same, it actually behaves differently depending on the `KernelInstall` flag.

**Comment by alfonsosanchezbeato** [comment]
I've now changed the name of the tasks. For the moment I have not changed the name of the methods (m.doSetupKernelSnap, m.undoSetupKernelSnap, etc.) to avoid too much noise in the PR, instead I would do that in a follow-up.

**Comment by alfonsosanchezbeato** [comment]
"Blocked" label removed now as we discussed this can be reviewed after changing the name of the tasks.

**Review by pedronis — COMMENTED** [review]
thanks, bunch of questions about the signatures and what is getting passed to the new helpers in setup.go?

File: overlord/snapstate/backend/setup.go [hunk 1]
```diff
@@ -325,3 +327,165 @@ func (b Backend) UndoSetupComponent(cpi snap.ContainerPlaceInfo, installRecord *
...
+		replaced, compsToRemove, ksnapName, ksnapRev, sysd,
+		"after failure to remove set-up of kernel modules components")
+}
+
+// mergeCompSideInfos returns a merged list from two lists of
+// ComponentSideInfo, and a list with the intersection, using the criteria of
+// the elements having the same ComponentRef. The rest of the data for an
+// element will come from comps2 if ComponentRef if the same in comps1 and
+// comps2.
+func mergeCompSideInfos(comps1, comps2 []*snap.ComponentSideInfo) (merged, instersect []*snap.ComponentSideInfo) {
```
Comment (pedronis) [inline overlord/snapstate/backend/setup.go:361]: "merge" is slightly confusing as the operation is not symmetric, as comps2 wins. We sometimes call this kind of functions "Delta" but maybe is doesn't clearly apply here?

File: overlord/snapstate/component.go [hunk 2]
```diff
@@ -169,6 +169,12 @@ func doInstallComponent(st *state.State, snapst *SnapState, compSetup *Component
 	}
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/component.go]: Hmm... maybe I'm missing something, but I'd say that probably this function is fine for component refreshes as well. For instance, I have tested that I can re-install components using `--dangerous`.

File: overlord/snapstate/snapmgr.go [hunk 3]
```diff
@@ -681,6 +681,7 @@ func Manager(st *state.State, runner *state.TaskRunner) (*SnapManager, error) {
 	runner.AddHandler("mount-component", m.doMountComponent, m.undoMountComponent)
 	runner.AddHandler("unlink-current-component", m.doUnlinkCurrentComponent, m.undoUnlinkCurrentComponent)
 	runner.AddHandler("link-component", m.doLinkComponent, m.undoLinkComponent)
+	runner.AddHandler("prepare-kernel-modules-components", m.doSetupKernelModules, m.doRemoveKernelModulesSetup)
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapmgr.go:684]: `testRemoveKernelModulesSetup` in `handlers_components_kernel_test.go` uses a task to trigger an error so both `doSetupKernelModules` and `doRemoveKernelModulesSetup` are called

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks for the review, I've pushed changes/answered questions now

File: overlord/snapstate/component.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/component.go]: but I thought we don't or can't use prepare-kernel-modules-components as is in a refresh?

File: overlord/snapstate/handlers_components.go [hunk 4]
```diff
@@ -396,3 +402,70 @@ func (m *SnapManager) undoUnlinkCurrentComponent(t *state.Task, _ *tomb.Tomb) (e
...
+	defer st.Unlock()
+	t.SetStatus(state.DoneStatus)
+	return nil
+}
+
+func (m *SnapManager) doRemoveKernelModulesSetup(t *state.Task, _ *tomb.Tomb) error {
+	// invariant: component unlinked on undo
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
```
Comment (pedronis) [inline overlord/snapstate/handlers_components.go:444]: should this one also follow the same style of locking as doSetup... ?

**Review by pedronis — APPROVED** [review]
thanks, especially for the added comments on behavior. one comment

File: overlord/snapstate/component.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/component.go]: Main split here is between the tasks that involve a kernel install/refresh (so a reboot is needed) vs when we update only kernel-modules components - I think that `prepare-kernel-modules-components` would be fine for both installing an updating for the latter case. But we will see if some problem arises.

File: overlord/snapstate/handlers_components.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/handlers_components.go]: Done now

File: overlord/snapstate/backend/setup.go [hunk 5]
```diff
@@ -325,3 +327,168 @@ func (b Backend) UndoSetupComponent(cpi snap.ContainerPlaceInfo, installRecord *
...
+// the elements having the same ComponentRef. The rest of the data for an
+// element will come from comps2 if ComponentRef is the same in comps1 and
+// comps2, that is, the revision is updated in that case.
+func mergeCompSideInfosUpdatingRev(comps1, comps2 []*snap.ComponentSideInfo) (merged, instersect []*snap.ComponentSideInfo) {
+	numInComps2 := len(comps2)
+	comps2Map := make(map[naming.ComponentRef]*snap.ComponentSideInfo, numInComps2)
+	for _, cti := range comps2 {
+		comps2Map[cti.Component] = cti
+	}
+	merged = append([]*snap.ComponentSideInfo(nil), comps2...)
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/setup.go]: Nice suggestion, done now

File: overlord/snapstate/backend/setup.go [hunk 6]
```diff
@@ -325,3 +327,168 @@ func (b Backend) UndoSetupComponent(cpi snap.ContainerPlaceInfo, installRecord *
...
+		compName, compRev, ksnapName, ksnapRev)
+
+	addMountUnitOptions := &systemd.MountUnitOptions{
+		MountUnitType: systemd.BeforeDriversLoadMountUnit,
+		Lifetime:      systemd.Persistent,
+		Description:   "Early mount for kernel-modules component",
+		What:          cpi.MountFile(),
+		Where:         dirs.StripRootDir(earlyMountDir),
+		Fstype:        "squashfs",
+		Options:       []string{"nodev,ro,x-gdu.hide,x-gvfs-hide"},
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/setup.go]: Good point, I now created the options in a method that leverages `squashfs.StandardOptions()`

File: overlord/snapstate/snapmgr.go [hunk 7]
```diff
@@ -643,8 +643,8 @@ func Manager(st *state.State, runner *state.TaskRunner) (*SnapManager, error) {
 	runner.AddHandler("conditional-auto-refresh", m.doConditionalAutoRefresh, nil)
 
 	// specific set-up for the kernel snap
-	runner.AddHandler("setup-kernel-snap", m.doSetupKernelSnap, m.undoSetupKernelSnap)
-	runner.AddHandler("remove-old-kernel-snap-setup", m.doCleanupOldKernelSnap, m.undoCleanupOldKernelSnap)
+	runner.AddHandler("prepare-kernel-snap", m.doSetupKernelSnap, m.undoSetupKernelSnap)
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapmgr.go:646]: This is not in any snapd release yet fortunately

File: overlord/snapstate/snapstate.go [hunk 8]
```diff
@@ -551,7 +551,7 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 
 	// This task is necessary only for UC20+ and hybrid
 	if deviceCtx.HasModeenv() && snapsup.Type == snap.TypeKernel {
-		setupKernel := st.NewTask("setup-kernel-snap", fmt.Sprintf(i18n.G("Setup kernel driver tree for %q%s"), snapsup.InstanceName(), revisionStr))
+		setupKernel := st.NewTask("prepare-kernel-snap", fmt.Sprintf(i18n.G("Prepare kernel driver tree for %q%s"), snapsup.InstanceName(), revisionStr))
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go:554]: Right, fixed already

**Review by alfonsosanchezbeato — COMMENTED** [review]
@bboozzoo thanks for the review, comments addressed now

**Review by bboozzoo — APPROVED** [review]
LGTM

**Comment by alfonsosanchezbeato** [comment]
History cleaned-up and branch re-pushed

**Comment by alfonsosanchezbeato** [comment]
Closing in favor of https://github.com/snapcore/snapd/pull/13748

**Pull request closed without merging**